### ISSUE TITLE
New (false positive?) InvalidReturnStatement in Psalm 5.5.0

### ISSUE THREAD

**Issue by BenMorel** [vimeo/psalm#9165]
The following code was working fine in previous versions:

```php
/**
 * @template T
 *
 * @template-implements IteratorAggregate<int, T>
 */
final class ImmutableList implements IteratorAggregate
{
    /**
     * @var list<T>
     */
    protected array $items;

    /**
     * @param list<T> $items
     */
    protected function __construct(array $items = [])
    {
        $this->items = $items;
    }

    /**
     * @return ArrayIterator<int, T>
     */
    public function getIterator(): ArrayIterator
    {
        return new ArrayIterator($this->items);
    }
}
```

But now fails with:

> ERROR: [InvalidReturnStatement](https://psalm.dev/128) - 27:16 - The inferred type 'ArrayIterator<int<0, max>, T:ImmutableList as mixed>' does not match the declared return type 'ArrayIterator<int, T:ImmutableList as mixed>' for ImmutableList::getIterator

> ERROR: [InvalidReturnType](https://psalm.dev/011) - 23:16 - The declared return type 'ArrayIterator<int, T:ImmutableList as mixed>' for ImmutableList::getIterator is incorrect, got 'ArrayIterator<int<0, max>, T:ImmutableList as mixed>'

https://psalm.dev/r/785517d895

If I change the return type of `getIterator()` to `ArrayIterator<int<0, max>, T>`, it works fine.

But as far as I understand it, it should work find with `<int, T>` as well? Or do I miss something?

**Comment by weirdan** [comment]
> But as far as I understand it, it should work find with `<int, T>` as well? 

It depends on whether that first generic parameter of `ArrayIterator` is covariant or not. The error is valid if it's not covariant.
